keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -183,90 +191,83 @@ open class SqlResourceRepository(
     }
   }
 
-  override fun eventHistory(id: String, limit: Int): List<ResourceEvent> {
+  override fun eventHistory(id: String, limit: Int): List<ResourceHistoryEvent> {
     require(limit > 0) { "limit must be a positive integer" }
+
+    val resource = get(id)
+
     return sqlRetry.withRetry(READ) {
       jooq
         .select(EVENT.JSON)
-        .from(EVENT, RESOURCE)
-        .where(EVENT.SCOPE.eq(Scope.RESOURCE.name))
-        .and(RESOURCE.ID.eq(id))
-        .and(RESOURCE.UID.eq(EVENT.UID))
+        .from(EVENT)
+        // look for resource events that match the resource...
+        .where(EVENT.SCOPE.eq(Scope.RESOURCE.name)
+          .and(EVENT.UID.eq(resource.uid))
+        )
+        // ...or application events that match the application as they apply to all resources
+        .or(EVENT.SCOPE.eq(Scope.APPLICATION.name)
+          .and(EVENT_JSON_APPLICATION.eq(resource.application))
+        )
         .orderBy(EVENT.TIMESTAMP.desc())
         .limit(limit)
         .fetch()
         .map { (json) ->
-          objectMapper.readValue<ResourceEvent>(json)
-        }
-        .ifEmpty {
-          throw NoSuchResourceId(id)
+          objectMapper.readValue<PersistentEvent>(json)
         }
+        // filter out application events that don't affect resource history
+        .filterIsInstance<ResourceHistoryEvent>()
     }
   }
 
-  override fun appendHistory(event: ApplicationEvent) {
-    jooq.transaction { config ->
-      val txn = DSL.using(config)
-
-      if (event.ignoreRepeatedInHistory) {
-        val previousEvent = txn
-          .select(EVENT.JSON)
-          .from(EVENT)
-          .where(EVENT.SCOPE.eq(Scope.APPLICATION.name))
-          .and(EVENT.UID.eq(event.application))
-          .orderBy(EVENT.TIMESTAMP.desc())
-          .limit(1)
-          .fetchOne()
-          ?.let { (json) ->
-            objectMapper.readValue<ApplicationEvent>(json)
-          }
-
-        if (event.javaClass == previousEvent?.javaClass) return@transaction
-      }
+  // todo: add sql retries once we've rethought repository structure: https://github.com/spinnaker/keel/issues/740
```

Comment:
@emjburns I kept this comment here because I didn't quite understand the reasoning behind it, but the issue linked is closed. I don't understand why we can't add retries here and elsewhere in this class. Could you please explain?

Reply:
tbh I don't remember why we couldn't do sql retries here. maybe it's not true!